library/core/src/num/mod.rs

```diff
@@ -846,37 +866,42 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+
+    if intrinsics::likely(
+        radix <= 16 && digits.len() <= mem::size_of::<T>() * 2 - is_signed_ty as usize,
+    ) {
+        // SAFETY: We can take this fast path when `radix.pow(digits.len()) - 1 <= T::MAX`
+        // but the condition above is a faster (conservative) approximation of this.
```

Comment:
```suggestion
        // SAFETY: We could take this fast path when `radix.pow(digits.len()) - 1 <= T::MAX`
        // but the condition above is a faster (conservative) approximation that covers the
        // common case (radix 10 or 16, and relatively small numbers).
```

Reply:
> and relatively small numbers

It's not exactly relatively small numbers.

Reply:
For `u64` the fast path is taken up to a length of 16 digits. For radix 10 that's still four orders of magnitude below `u64::MAX`. At least to me that qualifies as "relatively small". For `i8` it's only singel digit numbers, which to me qualify as "small" without "relatively".

Feel free to suggest an alternative wording though.

Reply:
It's more than half of `u64` which is why I should say it should be "relatively large". Maybe we shouldn't use the word "relatively" at all, people may misinterpret it differently. Maybe we should say the "whichever number nearer to the binary representation with minimum operations while covering a large set of numbers".

Reply:
But that's like saying numbers below 10 are a "relatively large" fraction of numbers up to 128. For `u64` and radix 10 this fast path is taken for less than 0.06% of representable numbers.
`(9.999.999.999.999.999 / u64::MAX == 0.000542101)`

> "whichever number nearer to the binary representation with minimum operations while covering a large set of numbers"
No clue what you're trying to say here, sorry.

Reply:
Words are harder than code :-)

Maybe we could write:

but the condition above is a fast approximation that is perfect for radix 16, ok for radix 10 and way too concervative for radix 2.

Reply:
s/concervative/conservative/
> perfect for radix 16

`i8` would like a word with you but the direction is good.

Maybe:
"but the condition above is a quicker to compute approximation that - when the target type is unsigned -  is perfect for radix 16, ok for radix 10 and way too concervative for radix 2. Signed types have three bits left unconsidered."


> Words are harder than code :-)

Oh, how true :smile:

Reply:
> But that's like saying numbers below 10 are a "relatively large" fraction of numbers up to 128. For u64 and radix 10 this fast path is taken for less than 0.06% of representable numbers.

I mean it's larger than `u32` in the case of `u64`, not `u64 / 2`.

Reply:
> 3 bits?
A radix 16 digit takes four bits to represent, but one bit is needed to represent the sign. The approximation therefore reserves an entire four bit block, meaning 3 bits are "wasted" ("unconsidered" was the wrong word).

Not sure what's the best way to express that in the comment though. Or if it should be in the comment at all.

What about:
"but the condition above is a quicker to compute approximation that - when the target type is unsigned - is perfect for radix 16, ok for radix 10 and very restrictive for radix 2. Signed types are restricted further by the need to represent the sign. That ends up reserving a full four bit block, since a radix 16 digit takes up four bit in the worst case."

Reply:
I realised I was going into the weeds to early and not explaining why we were trying to do it. I've tried to do a high level `why` and expanded the example to look at u8 and i8 as I think that makes it pretty explicit as to what's going on.

Reply:
@LingMan are you happy with the new wording?

Reply:
Hm, I don't see a new wording. Did you push?

Reply:
I did notice this today: "Your branch is ahead of 'origin/plan_b' by 1 commit."
D'oh, sorry @LingMan

Reply:
Opened a new sub-thread, since this one's now marked outdated - and sorry for the late reply.